replica/database.cc

```diff
@@ -1842,13 +1842,19 @@ future<> database::apply(schema_ptr s, const frozen_mutation& m, tracing::trace_
         update_write_metrics_for_timed_out_write();
         return make_exception_future<>(timed_out_error{});
     }
+    if (!s->is_synced()) {
+        on_internal_error(dblog, format("attempted to apply mutation using not synced schema of {}.{}, version={}", s->ks_name(), s->cf_name(), s->version()));
+    }
     return update_write_metrics(_apply_stage(this, std::move(s), seastar::cref(m), std::move(tr_state), timeout, sync));
```

Comment:
The changelog says the current check is too late, but doesn't explain why.

Shouldn't we move the check rather than add one?

Reply:
Throwing the error from `do_apply` causes the query to fail but it is too late since it doesn't help find the offending path.
We better get a stack trace before the apply stage is engaged so we know what path was using an unsync'ed schema.

And yes, the check can be moved up the stack since all paths leading to `_apply_stage` / `do_apply` (which are private) are covered.